src/appservice/config/config.ts

```diff
@@ -8,44 +8,117 @@
-  webAPI: {
-    port: number;
-  };
-  /** The admin room for the appservice bot. Not called managementRoom like draupnir on purpose, so they're not mixed in code somehow. */
-  adminRoom: string;
-  /** configuration for matrix-appservice-bridge's Logger */
-  logging?: LoggingOpts;
-
-  dataPath: string;
-
-  // Store room state using sqlite to improve startup time when Synapse responds
-  // slowly to requests for `/state`.
-  roomStateBackingStore: {
-    enabled?: boolean;
-  };
-}
-
-export function read(configPath: string): IConfig {
+export function read(configPath: string): AppserviceConfig {
   const content = fs.readFileSync(configPath, "utf8");
-  const parsed = load(content);
-  const config = parsed as object as IConfig;
-  return config;
+  const jsonParsed = load(content);
+  const decodedConfig = Value.Decode(AppserviceConfig, jsonParsed);
+  return decodedConfig;
 }
+
+export const LoggingOptsSchema = Type.Object({
+  console: Type.Optional(
+    Type.Union(
+      [
+        Type.Literal("debug"),
+        Type.Literal("info"),
+        Type.Literal("warn"),
+        Type.Literal("error"),
+        Type.Literal("trace"),
+        Type.Literal("off"),
+      ],
+      { description: "The log level used by the console output." }
+    )
+  ),
+  json: Type.Optional(
+    Type.Boolean({
+      description:
+        "Should the logs be outputted in JSON format, for consumption by a collector.",
+    })
+  ),
+  colorize: Type.Optional(
+    Type.Boolean({
+      description:
+        "Should the logs color-code the level strings in the output.",
+    })
+  ),
+  timestampFormat: Type.Optional(
+    Type.String({
+      description: "Timestamp format used in the log output.",
+      default: "HH:mm:ss:SSS",
+    })
+  ),
+});
+
+export type AppserviceConfig = EDStatic<typeof AppserviceConfig>;
+export const AppserviceConfig = Type.Object({
+  homeserver: Type.Object(
+    {
+      domain: Type.String({
+        description:
+          "The domain of the homeserver that is found at the end of mxids",
+      }),
+      url: Type.String({
+        description:
+          "The url to use to access the client server api e.g. 'https://matrix-client.matrix.org'",
+      }),
+    },
+    {
+      description: "Details for the homeserver the appservice will be serving ",
+    }
+  ),
+  db: Type.Object(
+    {
+      connectionString: Type.String({
+        description: "Postgres connection string",
+      }),
+    },
+    { description: "Details for the database backend" }
+  ),
+  webAPI: Type.Object(
+    {
+      port: Type.Number({
+        description:
+          "Port number for the web API used to access the appservice via the widget",
+      }),
+    },
+    {
+      description:
+        "Config for the web api used to access the appservice via the widget",
+    }
+  ),
+  adminRoom: Type.Union(
+    [StringRoomIDSchema, StringRoomAliasSchema, PermalinkSchema],
```

Comment:
This won't do, we need a `MatrixRoomReference` transform instead of this union because the PermalinkSchema will always fail when a roomID or alias is provided... no i don't know why this isn't failing in integration tests. Needs investigating also.